File: bsearch/go/bsearch.go
package bsearch

func bsearch(arr []int, target int) int {
    for l, r := 0, len(arr)-1; l <= r; {
        mid := l + ((r - l) / 2)
        // target less than middle, so go left side
        if target < arr[mid] {
            r = mid-1
            continue
        // target greater than middle, so go right side
        } else if target > arr[mid] {
            l = mid+1
            continue
        }
        // target equals middle, return it
        return mid
    }
    // target does not exist in the arr
    return -1
}

File: bsearch/c/source/bsearch.c
#include "bsearch.h"
#include <stddef.h>
#include <stdlib.h>

int
_bsearch(int const *arr, size_t n, int target) {
	size_t l,r;
	l = 0;
	r = n;
	while (l <= r) {
		size_t mid = l + ((r - l)/2);
		if (target < arr[mid]) { // search in the fleft side
			r = mid-1;
			continue;
		} else if (target > arr[mid]) { // search in the right side
			l = mid+1;
			continue;
		}
		return mid;
	}
	return -1;
}
